File: Mage/ObservationAnnotationView.m
//
//  ObservationAnnotationView.m
//  MAGE
//
//  Created by William Newman on 1/19/16.
//

#import "ObservationAnnotationView.h"

@interface ObservationAnnotationView ()

@property (nonatomic, strong) MKMapView *mapView;
@property (nonatomic, strong) NSObject<AnnotationDragCallback> *dragCallback;

@end

@implementation ObservationAnnotationView

- (instancetype)initWithAnnotation:(id <MKAnnotation>)annotation reuseIdentifier:(NSString *)reuseIdentifier andMapView: (MKMapView *) mapView andDragCallback: (NSObject<AnnotationDragCallback> *) dragCallback{
    self = [super initWithAnnotation:annotation reuseIdentifier:reuseIdentifier];
    if(self){
        self.mapView = mapView;
        self.dragCallback = dragCallback;
    }
    return self;
}

- (void)setDragState:(MKAnnotationViewDragState)newDragState animated:(BOOL)animated {
    [super setDragState:newDragState animated:animated];

    if (newDragState == MKAnnotationViewDragStateStarting) {
        __weak __typeof__(self) weakSelf = self;
        [UIView animateWithDuration:0.3 animations:^{
            CGRect imageFrame = weakSelf.frame;
            imageFrame.origin.y = imageFrame.origin.y - weakSelf.image.size.height;
            [weakSelf setFrame:imageFrame];
         }
         completion:^(BOOL finished) {
             weakSelf.dragState = MKAnnotationViewDragStateDragging;
         }];
    
    } else if (newDragState == MKAnnotationViewDragStateEnding) {
        __weak __typeof__(self) weakSelf = self;
        [UIView animateWithDuration:.2 animations:^{
            CGRect imageFrame = weakSelf.frame;
            imageFrame.origin.y = (imageFrame.origin.y - (weakSelf.image.size.height / 2));
            [weakSelf setFrame:imageFrame];
         }
         completion:^(BOOL finished) {
             [UIView animateWithDuration:.2 animations:^{
                 CGRect imageFrame = weakSelf.frame;
                 imageFrame.origin.y = imageFrame.origin.y + (weakSelf.image.size.height / 2);
                 [weakSelf setFrame:imageFrame];
              }
              completion:^(BOOL finished) {
                  weakSelf.dragState = MKAnnotationViewDragStateNone;
              }];
         }];
    } else if (newDragState == MKAnnotationViewDragStateCanceling) {
        [UIView animateWithDuration:0.2 animations:^{
            CGRect imageFrame = self.frame;
            imageFrame.origin.y = imageFrame.origin.y + self.image.size.height;
            [self setFrame:imageFrame];
         }
         completion:^(BOOL finished) {
             self.dragState = MKAnnotationViewDragStateNone;
         }];
    }
}

- (void) setCenter:(CGPoint)center{
    [super setCenter:center];
    if(self.dragCallback != nil){
        CGPoint point = CGPointMake(center.x, center.y - self.centerOffset.y);
        CLLocationCoordinate2D coordinate = [_mapView convertPoint:point toCoordinateFromView:self.superview];
        [self.dragCallback draggingAnnotationView:self atCoordinate:coordinate];
    }
}

@end

File: Mage/ObservationAnnotationView.swift
//
//  ObservationAnnotationView.m
//  MAGE
//
//  Created by William Newman on 1/19/16.
//

import Foundation
import MapKit

@objc class ObservationAnnotationView : MKAnnotationView {
    
    var mapView: MKMapView
    var dragCallback: AnnotationDragCallback?
    
    @objc public init(annotation: MKAnnotation?, reuseIdentifier: String, mapView: MKMapView, dragCallback: AnnotationDragCallback?) {
        self.mapView = mapView
        self.dragCallback = dragCallback
        super.init(annotation: annotation, reuseIdentifier: reuseIdentifier)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override var center: CGPoint {
        set {
            super.center = newValue
            if let dragCallback = dragCallback {
                let point = CGPoint(x: newValue.x, y: newValue.y - centerOffset.y)
                let coordinate = mapView.convert(point, toCoordinateFrom: superview)
                dragCallback.dragging(self, at: coordinate)
            }
        }
        get {
            return super.center
        }
    }
    
    override func setDragState(_ newDragState: MKAnnotationView.DragState, animated: Bool) {
        super.setDragState(newDragState, animated: animated)
        if newDragState == .starting {
            UIView.animate(withDuration: 0.3) {
                var imageFrame = self.frame
                imageFrame.origin.y = imageFrame.origin.y - (self.image?.size.height ?? 0.0)
                self.frame = imageFrame
            } completion: { finished in
                self.dragState = .dragging
            }
        } else if newDragState == .ending {
            UIView.animate(withDuration: 0.2) {
                var imageFrame = self.frame
                imageFrame.origin.y = imageFrame.origin.y - ((self.image?.size.height ?? 0.0) / 2)
                self.frame = imageFrame
            } completion: { finished in
                UIView.animate(withDuration: 0.2) {
                    var imageFrame = self.frame
                    imageFrame.origin.y = imageFrame.origin.y + ((self.image?.size.height ?? 0.0) / 2)
                    self.frame = imageFrame
                } completion: { finished in
                    self.dragState = .none
                }
            }
        } else if newDragState == .canceling {
            UIView.animate(withDuration: 0.2) {
                var imageFrame = self.frame
                imageFrame.origin.y = imageFrame.origin.y + (self.image?.size.height ?? 0.0)
                self.frame = imageFrame
            } completion: { finished in
                self.dragState = .none
            }
        }
    }
}
